lib/jsdom/web-idl/implementation/EventTarget.js

```diff
@@ -0,0 +1,231 @@
+    } else if (typeof callback === "object") {
+      callback = callback.handleEvent;
+    } else if (typeof callback !== "function") {
+      throw new TypeError("Only undefined, null, an object, or a function are allowed for the callback parameter");
+    }
+
+    capture = Boolean(capture);
+
+    if (callback === null) {
+      return;
+    }
+
+    if (this._events[type] === undefined) {
+      this._events[type] = {
+        capturing: [],
+        bubbling: []
+      };
+    }
+
+    var handlers = this._events[type][capture ? "capturing" : "bubbling"];
+    if (handlers.indexOf(callback) !== -1) {
+      return;
+    }
+
+    handlers.unshift(callback);
+  }
+
+  removeEventListener(type, callback, capture) {
+    if (callback === undefined || callback === null) {
+      callback = null;
+    } else if (typeof callback === "object") {
+      callback = callback.handleEvent;
+    } else if (typeof callback !== "function") {
+      throw new TypeError("Only undefined, null, an object, or a function are allowed for the callback parameter");
+    }
+
+    capture = Boolean(capture);
+
+    if (callback === null) {
+      // Optimization, not in the spec.
+      return;
+    }
+
+    if (this._events[type] === undefined) {
+      return;
+    }
+
+    var handlers = this._events[type][capture ? "capturing" : "bubbling"];
+    var idx = handlers.indexOf(callback);
+
+    if (idx !== -1) {
+      handlers.splice(idx, 1);
+    }
+  }
+
+  dispatchEvent(event) {
+    if (!("_bubbles" in event) || !("_cancelable" in event)) {
+      throw new TypeError("Argument to dispatchEvent must be an Event");
+    }
+
+    if (event._type === null || event._type === "") {
+      throw new DOMException(DOMException.INVALID_STATE_ERR, "Tried to dispatch an uninitialized event");
+    }
+
+    var targetList = [];
+
+    event._target = this;
+
+    //per the spec we gather the list of targets first to ensure
+    //against dom modifications during actual event dispatch
+    var target = this;
+    var targetParent = domSymbolTree.parent(target);
+    while (targetParent) {
+      targetList.push(targetParent);
+      target = targetParent;
+      targetParent = domSymbolTree.parent(target);
+    }
+    targetParent = target.defaultView;
+    if (targetParent) {
+      targetList.push(targetParent);
+    }
+
+    var iterator = backwardIterator(targetList);
+
+    event._eventPhase = event.CAPTURING_PHASE;
+    dispatchPhase(event, iterator);
+
+    iterator = singleIterator(event._target);
+    event._eventPhase = event.AT_TARGET;
+    dispatchPhase(event, iterator);
+
+    if (event._bubbles) {
+      iterator = forwardIterator(targetList);
+      event._eventPhase = event.BUBBLING_PHASE;
+      dispatchPhase(event, iterator);
+    }
+
+    event._currentTarget = null;
+    event._eventPhase = event.NONE;
+
+    return !event._canceled;
+  }
+}
+
+module.exports = {
+  init() {
+    this._events = Object.create(null);
+  },
+  interface: EventTarget
```

Comment:
Well I'm more just concerned that `interface` is an invalid name here. This isn't the interface; it's the implementation class behind the interface.